File: client-ios/Sources/Core/Stores/SettingsStore.swift
import Foundation

final class SettingsStore {
    private enum Key {
        static let host = "host"
        static let reconnectCid = "reconnect_cid"
        static let pushEndpoint = "push_endpoint"
        static let language = "language"
        static let defaultCameraEnabled = "default_camera_enabled"
        static let defaultMicEnabled = "default_mic_enabled"
        static let hdVideoExperimentalEnabled = "hd_video_experimental_enabled"
        static let savedRoomsShownFirst = "saved_rooms_shown_first"
        static let roomInviteNotificationsEnabled = "room_invite_notifications_enabled"
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = SettingsStore.defaultStore()) {
        self.defaults = defaults
    }

    var host: String {
        get {
            let value = defaults.string(forKey: Key.host)?.trimmingCharacters(in: .whitespacesAndNewlines)
            return value?.isEmpty == false ? value! : AppConstants.defaultHost
        }
        set {
            let normalized = newValue.trimmingCharacters(in: .whitespacesAndNewlines)
            if normalized.isEmpty { return }
            defaults.set(normalized, forKey: Key.host)
        }
    }

    var reconnectCid: String? {
        get {
            let value = defaults.string(forKey: Key.reconnectCid)?.trimmingCharacters(in: .whitespacesAndNewlines)
            return value?.isEmpty == false ? value : nil
        }
        set {
            if let value = newValue?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty {
                defaults.set(value, forKey: Key.reconnectCid)
            } else {
                defaults.removeObject(forKey: Key.reconnectCid)
            }
        }
    }

    var pushEndpoint: String? {
        get {
            let value = defaults.string(forKey: Key.pushEndpoint)?.trimmingCharacters(in: .whitespacesAndNewlines)
            return value?.isEmpty == false ? value : nil
        }
        set {
            if let value = newValue?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty {
                defaults.set(value, forKey: Key.pushEndpoint)
            } else {
                defaults.removeObject(forKey: Key.pushEndpoint)
            }
        }
    }

    var language: String {
        get {
            normalizeLanguage(defaults.string(forKey: Key.language))
        }
        set {
            defaults.set(normalizeLanguage(newValue), forKey: Key.language)
        }
    }

    var isDefaultCameraEnabled: Bool {
        get {
            if defaults.object(forKey: Key.defaultCameraEnabled) == nil { return true }
            return defaults.bool(forKey: Key.defaultCameraEnabled)
        }
        set {
            defaults.set(newValue, forKey: Key.defaultCameraEnabled)
        }
    }

    var isDefaultMicrophoneEnabled: Bool {
        get {
            if defaults.object(forKey: Key.defaultMicEnabled) == nil { return true }
            return defaults.bool(forKey: Key.defaultMicEnabled)
        }
        set {
            defaults.set(newValue, forKey: Key.defaultMicEnabled)
        }
    }

    var isHdVideoExperimentalEnabled: Bool {
        get {
            defaults.bool(forKey: Key.hdVideoExperimentalEnabled)
        }
        set {
            defaults.set(newValue, forKey: Key.hdVideoExperimentalEnabled)
        }
    }

    var areSavedRoomsShownFirst: Bool {
        get {
            defaults.bool(forKey: Key.savedRoomsShownFirst)
        }
        set {
            defaults.set(newValue, forKey: Key.savedRoomsShownFirst)
        }
    }

    var areRoomInviteNotificationsEnabled: Bool {
        get {
            if defaults.object(forKey: Key.roomInviteNotificationsEnabled) == nil { return true }
            return defaults.bool(forKey: Key.roomInviteNotificationsEnabled)
        }
        set {
            defaults.set(newValue, forKey: Key.roomInviteNotificationsEnabled)
        }
    }

    func normalizeLanguage(_ raw: String?) -> String {
        guard let raw else { return AppConstants.languageAuto }
        let value = raw.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        if AppConstants.supportedLanguages.contains(value) {
            return value
        }
        return AppConstants.languageAuto
    }

    private static func defaultStore() -> UserDefaults {
        UserDefaults(suiteName: AppConstants.appGroupIdentifier) ?? .standard
    }
}

File: client-android/app/src/main/java/app/serenada/android/data/SettingsStore.kt
package app.serenada.android.data

import android.content.Context
import android.content.SharedPreferences

class SettingsStore(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("serenada_settings", Context.MODE_PRIVATE)

    var host: String
        get() = prefs.getString(KEY_HOST, DEFAULT_HOST) ?: DEFAULT_HOST
        set(value) {
            val cleanValue = value.trim()
            if (cleanValue.isNotBlank()) {
                prefs.edit().putString(KEY_HOST, cleanValue).apply()
            }
        }

    var reconnectCid: String?
        get() = prefs.getString(KEY_RECONNECT_CID, null)
        set(value) {
            val editor = prefs.edit()
            if (value.isNullOrBlank()) {
                editor.remove(KEY_RECONNECT_CID)
            } else {
                editor.putString(KEY_RECONNECT_CID, value)
            }
            editor.apply()
        }

    var pushEndpoint: String?
        get() = prefs.getString(KEY_PUSH_ENDPOINT, null)
        set(value) {
            val editor = prefs.edit()
            val normalized = value?.trim()
            if (normalized.isNullOrBlank()) {
                editor.remove(KEY_PUSH_ENDPOINT)
            } else {
                editor.putString(KEY_PUSH_ENDPOINT, normalized)
            }
            editor.apply()
        }

    var language: String
        get() = normalizeLanguage(prefs.getString(KEY_LANGUAGE, LANGUAGE_AUTO))
        set(value) {
            prefs.edit().putString(KEY_LANGUAGE, normalizeLanguage(value)).apply()
        }

    var isDefaultCameraEnabled: Boolean
        get() = prefs.getBoolean(KEY_DEFAULT_CAMERA_ENABLED, true)
        set(value) {
            prefs.edit().putBoolean(KEY_DEFAULT_CAMERA_ENABLED, value).apply()
        }

    var isDefaultMicrophoneEnabled: Boolean
        get() = prefs.getBoolean(KEY_DEFAULT_MIC_ENABLED, true)
        set(value) {
            prefs.edit().putBoolean(KEY_DEFAULT_MIC_ENABLED, value).apply()
        }

    var isHdVideoExperimentalEnabled: Boolean
        get() = prefs.getBoolean(KEY_HD_VIDEO_EXPERIMENTAL_ENABLED, false)
        set(value) {
            prefs.edit().putBoolean(KEY_HD_VIDEO_EXPERIMENTAL_ENABLED, value).apply()
        }

    var areSavedRoomsShownFirst: Boolean
        get() = prefs.getBoolean(KEY_SAVED_ROOMS_SHOWN_FIRST, false)
        set(value) {
            prefs.edit().putBoolean(KEY_SAVED_ROOMS_SHOWN_FIRST, value).apply()
        }

    var areRoomInviteNotificationsEnabled: Boolean
        get() = prefs.getBoolean(KEY_ROOM_INVITE_NOTIFICATIONS_ENABLED, true)
        set(value) {
            prefs.edit().putBoolean(KEY_ROOM_INVITE_NOTIFICATIONS_ENABLED, value).apply()
        }


    companion object {
        const val DEFAULT_HOST = "serenada.app"
        const val HOST_RU = "serenada-app.ru"

        val PREDEFINED_HOSTS = listOf(DEFAULT_HOST, HOST_RU)

        const val LANGUAGE_AUTO = "auto"
        const val LANGUAGE_EN = "en"
        const val LANGUAGE_RU = "ru"
        const val LANGUAGE_ES = "es"
        const val LANGUAGE_FR = "fr"

        private const val KEY_HOST = "host"
        private const val KEY_RECONNECT_CID = "reconnect_cid"
        private const val KEY_PUSH_ENDPOINT = "push_endpoint"
        private const val KEY_LANGUAGE = "language"
        private const val KEY_DEFAULT_CAMERA_ENABLED = "default_camera_enabled"
        private const val KEY_DEFAULT_MIC_ENABLED = "default_mic_enabled"
        private const val KEY_HD_VIDEO_EXPERIMENTAL_ENABLED = "hd_video_experimental_enabled"
        private const val KEY_SAVED_ROOMS_SHOWN_FIRST = "saved_rooms_shown_first"
        private const val KEY_ROOM_INVITE_NOTIFICATIONS_ENABLED = "room_invite_notifications_enabled"

        fun normalizeLanguage(value: String?): String =
            when (value) {
                LANGUAGE_AUTO, LANGUAGE_EN, LANGUAGE_RU, LANGUAGE_ES, LANGUAGE_FR -> value
                else -> LANGUAGE_AUTO
            }
    }
}
